src/storage/src/source/generator/marketing.rs

```diff
@@ -0,0 +1,327 @@
+                    counter += 1;
+
+                    let mut lead = Lead {
+                        id,
+                        customer_id: rng.gen_range(0..CUSTOMERS.len()).try_into().unwrap(),
+                        created_at: now(),
+                        converted_at: None,
+                        conversion_amount: None,
+                    };
+
+                    pending.push((LEADS_OUTPUT, lead.to_row(), 1));
+
+                    // a highly scientific statistical model
+                    // predicting the likelyhood of a conversion
+                    let score = rng.sample::<f64, _>(Standard);
+                    let label = score > 0.5f64;
+
+                    let bucket = if lead.id % 10 <= 1 {
+                        CONTROL
+                    } else {
+                        EXPERIMENT
+                    };
+
+                    let mut prediction = Row::with_capacity(4);
+                    let mut packer = prediction.packer();
+
+                    packer.push(Datum::Int64(lead.id));
+                    packer.push(Datum::String(bucket));
+                    packer.push(Datum::TimestampTz(
+                        to_datetime(now()).try_into().expect("timestamp must fit"),
+                    ));
+                    packer.push(Datum::Float64(score.into()));
+
+                    pending.push((CONVERSIONS_PREDICTIONS_OUTPUT, prediction, 1));
+
+                    let mut sent_coupon = false;
+                    if !label && bucket == EXPERIMENT {
+                        sent_coupon = true;
+                        let amount = rng.gen_range(500..5000);
+
+                        let mut coupon = Row::with_capacity(4);
+                        let mut packer = coupon.packer();
+
+                        let id = counter;
+                        counter += 1;
+                        packer.push(Datum::Int64(id));
+                        packer.push(Datum::Int64(lead.id));
+                        packer.push(Datum::TimestampTz(
+                            to_datetime(now()).try_into().expect("timestamp must fit"),
+                        ));
+                        packer.push(Datum::Int64(amount));
+
+                        pending.push((COUPONS_OUTPUT, coupon, 1));
+                    }
+
+                    // Decide if a lead will convert. We assume our model is fairly
+                    // accurate and correlates with conversions. We also assume
+                    // that coupons make leads a little more liekly to convert.
+                    let mut converted = rng.sample::<f64, _>(Standard) < score;
+                    if sent_coupon && !converted {
+                        converted = rng.sample::<f64, _>(Standard) < score;
+                    }
+
+                    if converted {
+                        let converted_at = now() + rng.gen_range(1..30);
+
+                        future_updates.insert(converted_at, (LEADS_OUTPUT, lead.to_row(), -1));
```

Comment:
what happens if the source is restarted before we emit this future update, but after we emit the insertion? On restart, won't we either
1. not emit a retraction
2. emit a retraction with a different time
?

Reply:
I'm increasingly worried about this and think that 2 is what goes on, then we have a retraction and a negative total diff, which may cause various things to panic.

Reply:
Nevermind, I think it's 1 and we just don't emit a retraction for this. It's safe because now() is used to generate some row, and then maybe emits the retraction later on. Upon restart, it'd generate a new now and new row, without trying to make a retraction for any previous rows before the restart.